exec/java-exec/src/test/java/org/apache/drill/exec/physical/unit/TestEmptyBatchMiniPlan.java

```diff
@@ -0,0 +1,453 @@
+
+    BatchSchema expectedSchema = new SchemaBuilder()
+        .addNullable("a", TypeProtos.MinorType.BIGINT)
+        .addNullable("b", TypeProtos.MinorType.BIGINT)
+        .withSVMode(BatchSchema.SelectionVectorMode.NONE)
+        .build();
+
+    new MiniPlanTestBuilder()
+        .root(joinBatch)
+        .expectSchema(expectedSchema)
+        .expectZeroRow(true)
+        .go();
+  }
+
+  @Test
+  public void testLeftHashJoinRightEmpty() throws Exception {
+    List<String> leftJsonBatches = Lists.newArrayList(
+        "[{\"a\": 50, \"b\" : 10 }]");
+
+    RecordBatch leftScan = new JsonScanBuilder()
+        .jsonBatches(leftJsonBatches)
+        .columnsToRead("a", "b")
+        .build();
+
+    RecordBatch right = createEmptyBatchFromJson();
+
+    RecordBatch joinBatch = new PopBuilder()
+        .physicalOperator(new HashJoinPOP(null, null, Lists.newArrayList(joinCond("a", "EQUALS", "a2")), JoinRelType.LEFT))
+        .addInput(leftScan)
+        .addInput(right)
+        .build();
+
+    BatchSchema expectedSchema = new SchemaBuilder()
+        .addNullable("a", TypeProtos.MinorType.BIGINT)
+        .addNullable("b", TypeProtos.MinorType.BIGINT)
+        .withSVMode(BatchSchema.SelectionVectorMode.NONE)
+        .build();
+
+    new MiniPlanTestBuilder()
+        .root(joinBatch)
+        .expectSchema(expectedSchema)
+        .baselineValues(50L, 10L)
+        .go();
+  }
+
+  @Test
+  public void testUnionFilterAll() throws Exception {
+    List<String> leftJsonBatches = Lists.newArrayList(
+        "[{\"a\": 5, \"b\" : 1 }]");
+
+    List<String> rightJsonBatches = Lists.newArrayList(
+        "[{\"a\": 50, \"b\" : 10 }]");
+
+    RecordBatch leftScan = new JsonScanBuilder()
+        .jsonBatches(leftJsonBatches)
+        .columnsToRead("a", "b")
+        .build();
+
+    RecordBatch leftFilter = new PopBuilder()
+        .physicalOperator(new Filter(null, parseExpr("a < 0"), 1.0f))
+        .addInput(leftScan)
+        .build();
+
+    RecordBatch rightScan = new JsonScanBuilder()
+        .jsonBatches(rightJsonBatches)
+        .columnsToRead("a", "b")
+        .build();
+
+    RecordBatch rightFilter = new PopBuilder()
+        .physicalOperator(new Filter(null, parseExpr("a < 0"), 1.0f))
+        .addInput(rightScan)
+        .build();
+
+    RecordBatch batch = new PopBuilder()
+        .physicalOperator(new UnionAll(Collections.EMPTY_LIST)) // Children list is provided through RecordBatch
+        .addInput(leftFilter)
+        .addInput(rightFilter)
+        .build();
+
+    BatchSchema expectedSchema = new SchemaBuilder()
+        .addNullable("a", TypeProtos.MinorType.BIGINT)
+        .addNullable("b", TypeProtos.MinorType.BIGINT)
+        .withSVMode(BatchSchema.SelectionVectorMode.NONE)
+        .build();
+
+    new MiniPlanTestBuilder()
+        .root(batch)
+        .expectSchema(expectedSchema)
+        .expectZeroRow(true)
+        .go();
+  }
+
+  @Test
+  public void testOutputProjectEmpty() throws Exception {
```

Comment:
This test does a great job of handling a wide variety of expressions.

But, looking at the code under test, it does not exercise the following code paths:

* `SELECT *`
* `SELECT foo.*, bar.*`
* Error in the expression tree materializer
* Map column references: `myMap.innerMap.foo`
* Array references: `columns[5]`

Reply:
Add unit case for *, foo.*, bar.*, map and array reference.

I did not cover the case of error in expression tree materializer. If you have an example to produce error in tree materializer, maybe you can share it here. On the other hand, as explained above, the code path is an existing code block.